File: 11631 - Dark roads.java
/**
	UVa 11631 - Dark Roads
	by Rico Tiongson
	Submitted: October 21, 2013
	Accepted 1.558s Java
	O(nlogn) time	
*/
import java.util.*;
import java.io.*;
public class Main {
	static int n, m, total;
	static PriorityQueue<Edge> pq = new PriorityQueue<Edge>();
	
	static int[] f;
	static int find( int x ){
		if( f[x] == -1 ) return x;
		return f[x] = find( f[x] );
	}
	
	static int Kruskals(){
		f = new int[n];
		Arrays.fill( f, -1 );
		while( pq.size() != 0 ){
			Edge e = pq.poll();
			int a = find( e.i );
			int b = find( e.j );
			if( a != b ){
				f[a] = b;
				total -= e.w;
			}
		}
		return total;
	}
	public static void main( String args[] ) throws IOException {
		BufferedReader br = new BufferedReader( new InputStreamReader( System.in ) );
		PrintWriter pw = new PrintWriter( new BufferedWriter( new OutputStreamWriter( System.out ) ) );
		
		String in, s[];
		int a, b, c;
		
		while( (in=br.readLine()) != null ){
			if( in.equals("0 0") ) break;
			pq.clear();
			total = 0;
			s = in.split(" ");
			n = Integer.parseInt( s[0] );
			m = Integer.parseInt( s[1] );
			for( int i=0; i<m; ++i ){
				s = br.readLine().split(" ");
				a = Integer.parseInt( s[0] );
				b = Integer.parseInt( s[1] );
				c = Integer.parseInt( s[2] );
				pq.offer( new Edge( a, b, c ) );
				total += c;
			}
			pw.println( Kruskals() );
		}
		
		pw.close();
	}
}
class Edge implements Comparable<Edge> {
	public int i, j, w;
	public Edge( int I, int J, int W ){
		i = I;
		j = J;
		w = W;
	}
	public int compareTo( Edge E ){
		return w - E.w;
	}
}

File: 11631 - Dark Roads.cpp
/**
	UVa 11631 - Dark Roads
	by Rico Tiongson
	Submitted: May 1, 2013
	Accepted 0.782s C++
	O(logn) time
*/
#include<iostream>
#include<algorithm>
using namespace std;
int f[200015],m,sum;
int find(const int& x){
	return f[x]==x?x:f[x]=find(f[x]);
}
struct node{
	int x,y,w;
	node():x(0),y(0),w(0){}
	void in(){cin>>x>>y>>w;sum+=w;}
	bool operator<(const node& rhs)const{
		return w<rhs.w;
	}
} l[200015];
int main(){
	int n,a,b,i,d;
	while(cin >> m >> n,m||n){
		for(i=0;i<m;i++) f[i]=i;
		sum=0;
		for(d=0;d<n;d++) l[d].in();
		sort(l,l+n);
		d=0;
		for(i=m;d<n&&i>1;d++){
			a = find(l[d].x);
			b = find(l[d].y);
			if(a!=b){
				f[a] = b;
				sum-=l[d].w;
				i--;
			}
		}
		cout << sum << endl;
	}
}
